### PR TITLE
[LWG motion 27] P2548R6 copyable_function

### LINKED ISSUE #6313: [2023-06 LWG Motion 27] P2548R6 copyable_function
https://wiki.edg.com/pub/Wg21varna/StrawPolls/P2548R6.pdf

### PR DESCRIPTION
**Pull request by burblebee**
Fixes #6313.
Fixes cplusplus/papers#1275

### CODE REVIEW

**Comment by frederick-vs-ja** [comment]
This also fixes cplusplus/papers#1275.

File: source/utilities.tex [hunk 1]
```diff
@@ -13431,6 +13460,449 @@
...
+and \tcode{f} has no target object.
+\end{itemize}
+Otherwise, \tcode{*this} has a target object of type \tcode{VT}
+direct-non-list-initialized with \tcode{std::forward<F>(f)}.
+
+\pnum
+\throws
+Any exception thrown by the initialization of the target object.
+May throw \tcode{bad_alloc} unless \tcode{VT} is
+a function pointer or a specialization of \tcode{reference_wrapper}.
```
Comment (JohelEGP) [inline source/utilities.tex:13671]:
```suggestion
a pointer or a specialization of \tcode{reference_wrapper}.
```
This is what is used in the next two constructors.
This suggestion should be editorial given
the constraint of _`is-callable-from`_ and
the fact that only pointers that point to functions can be callable.

File: source/utilities.tex [hunk 2]
```diff
@@ -13431,6 +13460,449 @@
...
+\begin{itemdescr}
+\pnum
+\expects
+\tcode{*this} has a target object.
+
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+return @\placeholder{INVOKE}@<R>(static_cast<F @\placeholder{inv-quals}@>(f), std::forward<ArgTypes>(args)...);
```
Comment (JohelEGP) [inline source/utilities.tex:13864]:
```suggestion
return \placeholder{INVOKE}<R>(static_cast<F \placeholder{inv-quals}>(f), std::forward<ArgTypes>(args)...);
```

File: source/utilities.tex [hunk 3]
```diff
@@ -13431,6 +13460,449 @@
 \end{itemdescr}
 \indextext{function object!wrapper|)}
```
Comment (JohelEGP) [inline source/utilities.tex:13461]:
```suggestion
```

File: source/utilities.tex [hunk 4]
```diff
@@ -13431,6 +13460,449 @@
...
+}
+\end{codeblock}
+
+\pnum
+The \tcode{copyable_function} class template provides polymorphic wrappers
+that generalize the notion of a callable object\iref{func.def}.
+These wrappers can store, copy, move, and call arbitrary callable objects,
+given a call signature.
+Within this subclause, \exposid{call-args} is an argument pack
+with elements that have types \tcode{ArgTypes\&\&...} respectively.
```
Comment (JohelEGP) [inline source/utilities.tex:13540]: This is unused.
```suggestion
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/utilities.tex (hunk 1 above)
Comment (burblebee) [reply source/utilities.tex:13671]: A question for the authors, but func ptrs are special cased in the rules for rval refs, so... intended?

File: source/utilities.tex (hunk 2 above)
Comment (burblebee) [reply source/utilities.tex:13864]: That produces:
`return \placeholder{INVOKE}<R>(static_cast<F \placeholder{inv-quals}>(f), std::forward<ArgTypes>(args)...)`

Looks fine as is, no? reopen if you disagree.

File: source/utilities.tex (hunk 3 above)
Comment (burblebee) [reply source/utilities.tex:13461]: done, thanks!

File: source/utilities.tex (hunk 4 above)
Comment (burblebee) [reply source/utilities.tex:13540]: done in separate commit, thanks!

File: source/utilities.tex [hunk 5]
```diff
@@ -13431,6 +13460,449 @@
...
+  };
+}
+\end{codeblock}
+
+\pnum
+The \tcode{copyable_function} class template provides polymorphic wrappers
+that generalize the notion of a callable object\iref{func.def}.
+These wrappers can store, copy, move, and call arbitrary callable objects,
+given a call signature.
+Within this subclause, \exposid{call-args} is an argument pack
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/utilities.tex:13539]: Superseded by comment below.

**Comment by burblebee** [comment]
Builds are failing during OS installs; how to fix?

File: source/utilities.tex (hunk 2 above)
Comment (JohelEGP) [reply source/utilities.tex:13861]: You're right. This is within a `codeblock` environment, so `@`s are needed.

**Review by JohelEGP — APPROVED** [review]

File: source/utilities.tex [hunk 6]
```diff
@@ -12639,6 +12644,30 @@
 Subclause \ref{func.wrap} describes polymorphic wrapper classes that
 encapsulate arbitrary callable objects.
 
+\pnum
+Let \tcode{t} be an object of a type that is a specialization of
+\tcode{function}, \tcode{copyable_function}, or \tcode{move_only_function},
+such that the target object \tcode{x} of \tcode{t} has a type that
+is a specialization of
+\tcode{function}, \tcode{copyable_function}, or \tcode{move_only_function}.
```
Comment (tkoeppe) [inline source/utilities.tex:12652]: Does `function_ref` need to be mentioned here (@CaseyCarter, @MFHava)?

Comment (MFHava) [reply source/utilities.tex:12652]: As the intention was that there should be no double wrapping for "polymorphic function wrappers", I think `function_ref` should be mentioned here too.

Comment (jensmaurer) [reply source/utilities.tex:12652]: Such a change feels non-editorial to me.

Comment (tkoeppe) [reply source/utilities.tex:12652]: Should we open an LWG issue?

**Pull request merged**